### DISCUSSION TITLE
Can I not use BroadcastChannel in NextJS?

### DISCUSSION THREAD

**Discussion by florianwalther-private** [vercel/next.js#37838] [Help] [answered]
I'm converting my vanilla React app to NextJS right now. It seems that Next doesn't allow me to use BroadcastChannel anymore. I get this error: ` BroadcastChannel is not defined`.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

You need to have use something I call, guarded DOM access, which includes window access as well.

The pitfall here is that your code is running also on the server, and `BroadcastChannel` doesn't exist in global object exposed by Node.js, instead you gotta import it from `worker_threads` (at least in the versions I've used), and if you want to use it client side you gotta use `window.BroadcastChannel` again, making sure `window` is not `undefined`.

Check this snippet:

```tsx
import type { GetServerSidePropsContext, NextPage } from "next";
import { useEffect } from "react";

import { BroadcastChannel } from "worker_threads";

const channel = new BroadcastChannel("bar");

const Home: NextPage = () => {
  useEffect(() => {
    console.log(new window.BroadcastChannel("foo"));
  }, []);
  return <div>Home</div>;
};

export const getServerSideProps = (ctx: GetServerSidePropsContext) => {
  console.log({ channel });
  return { props: {} };
};

export default Home;
```

If we pass this through the [code-elimination-tool](https://next-code-elimination.vercel.app/s/Mxb_0mD8U):

```tsx
// This is the code that is bundled for the client-side:

import type { NextPage } from 'next';
import { useEffect } from 'react';
const Home: NextPage = () => {
  useEffect(() => {
    console.log(new window.BroadcastChannel('foo'));
  });
  return <div>Home</div>;
};
export var __N_SSP = true;
export default Home;
```

I hope this help, and that I was correct about `BrodcastChannel`'s presence in the Node.js, or lack thereof.

All in all, you have to be careful not to mix these two. This can happen in very sneaky ways, via imports, or unguarded access.

**Reply by florianwalther-private** [reply]
Thank you for your answer! I only need the BroadcastChannel on the client-side, so I wrapped it in ` const isBrowser = () => typeof window !== 'undefined';`. It works 👍

**Reply by icyJoseph** [reply]
Awesome, please don't forget to mark as answered!

**Reply by balazsorban44** [reply]
This answer is correct, thanks @icyJoseph! The documentation for this can be found here: https://nextjs.org/docs/migrating/from-create-react-app#safely-accessing-web-apis

Note that inside `useEffect`, you don't need to write `window.BroadcastChannel` either, just `BroadcastChannel`, as Next.js knows you have a different global context. :+1:
